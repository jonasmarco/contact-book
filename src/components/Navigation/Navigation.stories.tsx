import { Meta, StoryFn } from '@storybook/react'
import { MemoryRouter } from 'react-router-dom'
import styled, { css } from 'styled-components'

import Navigation from '.'

const Background = styled.div`
  ${({ theme }) => css`
    background-color: ${theme.colors.tertiary};
    height: 100vh;
    padding: 0;
    width: 100vw;
  `};
`

export default {
  title: 'Navigation',
  component: Navigation,
  decorators: [
    (Story) => (
      <MemoryRouter initialEntries={['/']}>
        <Background>
          <Story />
        </Background>
      </MemoryRouter>
    )
  ],
  parameters: {
    layout: 'fullscreen'
  }
} as Meta<typeof Navigation>

const Template: StoryFn<typeof Navigation> = () => <Navigation />

export const SimpleNavigation = Template.bind({})
