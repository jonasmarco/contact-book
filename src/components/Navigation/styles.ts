import styled, { css } from 'styled-components'
import { NavLink as Link } from 'react-router-dom'

export const Nav = styled.nav`
  ${({ theme }) => css`
    border-right: 0.1rem solid rgba(255, 255, 255, 0.1);
    box-shadow: inset 5px 0px 10px -8px rgb(0 0 0 / 12%);
    display: flex;
    height: 100vh;
    justify-content: center;
    padding-top: 6rem;
    width: 6rem;
    > ul {
      display: flex;
      flex-direction: column;
      width: 100%;
    }
  `}
`

export const MenuItem = styled(Link)`
  ${({ theme }) => css`
    align-items: center;
    border-bottom: 0.1rem solid rgba(255, 255, 255, 0.1);
    display: flex;
    height: 4rem;
    justify-content: center;
    svg {
      color: ${theme.colors.white};
      width: 1.8rem;
    }
    &:hover,
    &:focus {
      background-color: ${theme.colors.gray};
    }
    &.active {
      background-color: ${theme.colors.gray};
      border-left: 0.1rem solid ${theme.colors.primary};
    }
  `}
`
