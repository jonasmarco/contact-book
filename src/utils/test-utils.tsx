import { ReactElement } from 'react'
import { render } from '@testing-library/react'
import { BrowserRouter as Router } from 'react-router-dom'

import { ThemeProvider } from 'styled-components'
import theme from '../styles/theme'

const customRender = (ui: ReactElement) =>
  render(
    <Router>
      <ThemeProvider theme={theme}>{ui}</ThemeProvider>
    </Router>
  )

// re-export everything
export * from '@testing-library/react'

// override render method
export { customRender as render }
