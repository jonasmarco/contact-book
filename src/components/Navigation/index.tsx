import InlineList from '@/components/InlineList'

import { BookContacts } from '@styled-icons/fluentui-system-regular'
import { Gear } from '@styled-icons/evil'

import * as S from './styles'

const Navigation = () => {
  return (
    <S.Nav>
      <InlineList>
        <li>
          <S.MenuItem to="/" data-testid="book-contacts">
            <BookContacts />
          </S.MenuItem>
        </li>
        <li>
          <S.MenuItem to="/contacts/import-contacts" data-testid="gear">
            <Gear />
          </S.MenuItem>
        </li>
      </InlineList>
    </S.Nav>
  )
}

export default Navigation
