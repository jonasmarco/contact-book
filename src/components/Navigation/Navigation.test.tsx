import { render, screen } from '../../utils/test-utils'
import Navigation from '.'

describe('Navigation Component', () => {
  // (renderiza sem erros):
  // ensures that the Navigation is rendered without errors
  test('is rendered without error', () => {
    render(<Navigation />)
  })

  // (renderiza com a Navigation com dois links):
  // ensure Navigation is rendered with two links
  test('is rendered the correct list items', () => {
    render(<Navigation />)

    const menuItems = screen.getAllByRole('listitem')
    expect(menuItems.length).toBe(2)
  })

  // (renderiza com a lista de links corretos):
  // ensure Navigation is rendered with correct links
  test('is rendered correct navigation links', () => {
    render(<Navigation />)

    const bookContactsLink = screen.getByTestId('book-contacts')
    const gearLink = screen.getByTestId('gear')

    expect(bookContactsLink).toBeInTheDocument()
    expect(bookContactsLink).toHaveAttribute('href', '/')

    expect(gearLink).toBeInTheDocument()
    expect(gearLink).toHaveAttribute('href', '/contacts/import-contacts')
  })

  // (snapshot)
  test('matches snapshot', () => {
    const { asFragment } = render(<Navigation />)
    expect(asFragment()).toMatchSnapshot()
  })
})
