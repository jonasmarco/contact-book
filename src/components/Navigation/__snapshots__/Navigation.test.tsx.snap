// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation Component matches snapshot 1`] = `
<DocumentFragment>
  .c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  list-style: none;
}

.c3 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 {
  border-right: 0.1rem solid rgba(255,255,255,0.1);
  box-shadow: inset 5px 0px 10px -8px rgb(0 0 0 / 12%);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100vh;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding-top: 6rem;
  width: 6rem;
}

.c0 > ul {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
}

.c2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-bottom: 0.1rem solid rgba(255,255,255,0.1);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 4rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c2 svg {
  color: #FCFCFC;
  width: 1.8rem;
}

.c2:hover,
.c2:focus {
  background-color: #596E7D;
}

.c2.active {
  background-color: #596E7D;
  border-left: 0.1rem solid #4FB6FF;
}

<nav
    class="c0"
  >
    <ul
      class="c1"
    >
      <li>
        <a
          aria-current="page"
          class="c2 active"
          data-testid="book-contacts"
          href="/"
        >
          <svg
            aria-hidden="true"
            class="c3"
            fill="currentColor"
            focusable="false"
            viewBox="0 0 32 32"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M21 16.5c0-.83-.67-1.5-1.5-1.5h-7c-.83 0-1.5.67-1.5 1.5v.5c0 1.97 1.86 4 5 4 3.14 0 5-2.03 5-4v-.5Zm-2.25-5.25a2.75 2.75 0 1 0-5.5 0 2.75 2.75 0 0 0 5.5 0ZM25 26H7.09c.2.58.76 1 1.41 1H26a1 1 0 1 1 0 2H8.5A3.5 3.5 0 0 1 5 25.5v-19A3.5 3.5 0 0 1 8.5 3h15A3.5 3.5 0 0 1 27 6.5V24a2 2 0 0 1-2 2ZM7 6.5V24h18V6.5c0-.83-.67-1.5-1.5-1.5h-15C7.67 5 7 5.67 7 6.5Z"
            />
          </svg>
        </a>
      </li>
      <li>
        <a
          class="c2"
          data-testid="gear"
          href="/contacts/import-contacts"
        >
          <svg
            aria-hidden="true"
            class="c3"
            fill="currentColor"
            focusable="false"
            viewBox="0 0 50 50"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M25 34c-5 0-9-4-9-9s4-9 9-9 9 4 9 9-4 9-9 9zm0-16c-3.9 0-7 3.1-7 7s3.1 7 7 7 7-3.1 7-7-3.1-7-7-7z"
            />
            <path
              d="M27.7 44h-5.4l-1.5-4.6c-1-.3-2-.7-2.9-1.2l-4.4 2.2-3.8-3.8 2.2-4.4c-.5-.9-.9-1.9-1.2-2.9L6 27.7v-5.4l4.6-1.5c.3-1 .7-2 1.2-2.9l-2.2-4.4 3.8-3.8 4.4 2.2c.9-.5 1.9-.9 2.9-1.2L22.3 6h5.4l1.5 4.6c1 .3 2 .7 2.9 1.2l4.4-2.2 3.8 3.8-2.2 4.4c.5.9.9 1.9 1.2 2.9l4.6 1.5v5.4l-4.6 1.5c-.3 1-.7 2-1.2 2.9l2.2 4.4-3.8 3.8-4.4-2.2c-.9.5-1.9.9-2.9 1.2L27.7 44zm-4-2h2.6l1.4-4.3.5-.1c1.2-.3 2.3-.8 3.4-1.4l.5-.3 4 2 1.8-1.8-2-4 .3-.5c.6-1 1.1-2.2 1.4-3.4l.1-.5 4.3-1.4v-2.6l-4.3-1.4-.1-.5c-.3-1.2-.8-2.3-1.4-3.4l-.3-.5 2-4-1.8-1.8-4 2-.5-.3c-1.1-.6-2.2-1.1-3.4-1.4l-.5-.1L26.3 8h-2.6l-1.4 4.3-.5.1c-1.2.3-2.3.8-3.4 1.4l-.5.3-4-2-1.8 1.8 2 4-.3.5c-.6 1-1.1 2.2-1.4 3.4l-.1.5L8 23.7v2.6l4.3 1.4.1.5c.3 1.2.8 2.3 1.4 3.4l.3.5-2 4 1.8 1.8 4-2 .5.3c1.1.6 2.2 1.1 3.4 1.4l.5.1 1.4 4.3z"
            />
          </svg>
        </a>
      </li>
    </ul>
  </nav>
</DocumentFragment>
`;
